Api client lib

calavera opened the pull request:
:warning: Too big to review, go to my fork and navigate the api/client package :warning: 

I extracted a standalone client library from the cli that can be used outside of docker. In the process of doing that I end up removing a bunch of code from the cli.
## Why do you do this to us??!!

We should own the main go library to interact with docker. This makes the docker cli be the first consumer of this library.
## How did you do this??!

I extracted all the calls the client needs into an interface, you can see it here:
https://github.com/calavera/docker/blob/api_client_lib/api/client/client.go#L19

Functions are very well defined but there are a couple things to consider.
### I removed `cli.stream` and `cli.hijack` entirely.

After reading every single line in the client, I believe nobody knows how those two functions work and they made our code very complicated. You can compare the start command, no need for a diff:

Master: https://github.com/docker/docker/blob/master/api/client/start.go
My branch: https://github.com/calavera/docker/blob/api_client_lib/api/client/start.go

Every response in the new library that can be streamed returns an `io.ReadCloser`. It's up to the caller to implement streaming. This makes the intention of the caller much more clear. Compare the image create functions:

Master: https://github.com/docker/docker/blob/master/api/client/create.go#L53-L70
My branch: https://github.com/calavera/docker/blob/api_client_lib/api/client/create.go#L51-L63

did you know that the response of that function was always a json message that we stream in the client?
## Breaking changes

Of course there are always consequences. I introduced what can be considered as a breaking change.  In my opinion, the new behavior is more correct :smile_cat: 

Until now, `docker inspect` continued evaluating containers if you typed a wrong id, see this example:

```
$ docker run -d --name test alpine true
$ docker inspect this_container_does_not_exist test
```

The message about `this_container_does_not_exist` not existing would be followed by the test container output, a json payload of about 100 lines. This makes almost impossible to see the error.

With the new behavior, the cli stops execution as soon as it finds an error. This is a more predictable behavior. Users know sooner and more clear that there is a problem with the command they just executed.

The code in inspect has also been highly simplified. Compare both files:

Master: https://github.com/docker/docker/blob/master/api/client/inspect.go
My branch: https://github.com/calavera/docker/blob/api_client_lib/api/client/inspect.go
                   https://github.com/calavera/docker/blob/api_client_lib/api/client/inspect/inspector.go
## This is impossible to review

I'm afraid so too. I tried very hard to keep changes very well defined. But the best way to review this change might be to browse my branch directly, and to play with the cli.

The new client library is in https://github.com/calavera/docker/blob/api_client_lib/api/client/lib

calavera committed:
Extract API client struct as standalone client.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker commit with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker cp with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker image create with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker container create with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker diff with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker events with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker export with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker history with the standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker images with the standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker import with the standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker info with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker build with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker kill with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker load with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker login with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker logs with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker network with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker pause with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker port with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker ps with standanlone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement container rename with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker restart with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker remove with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker rmi with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker save with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker stop with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker tag with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker top with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement trusted tagging with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker unpause with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker version with standalone client lib.

Use Go template definitions for default format.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker wait with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement getExitCode with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Use ensureReaderClosed consistently to close a response body reader.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker volume with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Fix client lib errors documentation.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Create interface that clients that talk to the api must fulfill.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Change references to query values.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Lowercase http method functions.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker attach with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker start with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker run with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker exec with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker stats with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker pull with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker push with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker inspect with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker resize with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Implement docker search with standalone client lib.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Remove old http from the docker cli.

Everything has been ported to the client library :tada:

Signed-off-by: David Calavera <david.calavera@gmail.com>

LK4D4 commented:
I think probably inspect should fail if any of containers doesn't exist :)

icecrime commented:
Oh my :heart_eyes: That looks pretty impressive!

LK4D4 commented:
Everything failed :fireworks:

LK4D4 commented:
Haha, seems like tests fixing won't be easy

duglin commented:
Its kind of disturbing that the change in semantics you mentioned didn't result in a change to a testcase.  Aside from squashing this down :-)  can we get a new testcase for that?

btw - I really like the idea of a cli lib!

vdemeester commented:
:heart_eyes_cat: Will have to take a look but `\o/`.

tonistiigi commented on api/client/client.go:
```diff
@@ -3,3 +3,74 @@
+	CopyToContainer(options types.CopyToContainerOptions) error
+	Events(options types.EventsOptions) (io.ReadCloser, error)
+	ImageBuild(options types.ImageBuildOptions) (types.ImageBuildResponse, error)
+	ImageCreate(options types.ImageCreateOptions) (io.ReadCloser, error)
```
Do we need both `ImageCreate` and `ImagePull`? Their implementation looks the same and `ImageCreate` is only called by pull helper. Also `Parent` in `ImageCreate` looks misleading as it is still just a reference to pull(together with tag).

calavera replied:
we probably don't need both, it's something to consider. I agree that the implementation is very similar.

tonistiigi commented on api/client/client.go:
```diff
@@ -3,3 +3,74 @@
+	ContainerExecInspect(execID string) (types.ContainerExecInspect, error)
+	ContainerExecResize(execID string, height, width int) error
+	ContainerExecStart(execID string, config *types.ExecStartCheck) error
+	ContainerExport(containerID string) (io.ReadCloser, error)
```
Lots of places where `containerID`/`imageID` is used it is not actually ID but a reference/prefix or container name. I think we should be clear about it on API level.

calavera replied:
I know that. I personally like `containerID` as a reference more than `containerIDOrNameOrShortID`, I've tried to be consistent everywhere, but I'm open to suggestions.

tonistiigi replied:
How about just `container/image` or `containerRef/imageRef`?

calavera replied:
I like names without the `ID` suffix, but I'm going to be very consistent, which requires to change pretty much every new type in the library. So these are the variable names without it:

```
containerID => container
imageID     => image
networkID   => network
volumeID    => volume
execID      => exec
```

I'll make the necessary changes if we agree on those.

tonistiigi replied:
I think they should be only replaced if they have the general meaning of `any-ref-to-container` or `any-ref-to-image`. For example, if create/commit return only IDs they should keep the `ID` suffix. Can `execID` be anything else than ID?

tonistiigi commented on api/client/client.go:
```diff
@@ -3,3 +3,74 @@
+
+// apiClient is an interface that clients that talk with a docker server must implement.
+type apiClient interface {
+	ContainerAttach(options types.ContainerAttachOptions) (*types.HijackedResponse, error)
```
Does this have to return pointer? + few other places (ExecStartCheck, ExecConfig). I think if there is no strict requirement for it we should be consistent with other methods.

calavera replied:
I've changed this reference to be consistent.

tonistiigi commented on api/client/stats.go:
```diff
@@ -156,18 +148,13 @@ func (cli *DockerCli) CmdStats(args ...string) error {
-		if *all {
-			v.Set("all", "1")
+		options := types.ContainerListOptions{
+			All: true,
```
should still check for `all` flag

calavera replied:
good catch! thank you.

tonistiigi commented:
`docker cp` seems to be broken. I get `http: read on closed response body` when copying from container and `Error response from daemon: Untar re-exec error: exit status 1: output: unexpected EOF` when copying to container. May may need to add more tests if this is the case(maybe they only check API atm).

calavera commented:
@tonistiigi yep, there are still a few tests failing that I'm currently fixing. `docker cp` included.

tianon commented:
Definitely high-level +1! :heart:

jfrazelle commented:
nice love the description definitely helps :)

calavera committed:
Implement all inspect commands with the new inspector interface.

It makes the behavior completely consistent across commands.
It adds tests to check that execution stops when an element is not
found.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
@duglin I totally agree with you, I was very suspicious about not seen inspect test failing. I've added some in my last commit. All inspects commands are implemented exactly in the same way, so all of them behave alike :raised_hands: 

https://github.com/docker/docker/commit/1db5748a35eb984a2e52ddb04a13723af0e7cbf9

tiborvass commented:
@calavera don't really want to start a name bikeshedding, but I don't think it's a good idea to have a package be named `lib`. `dockerclient` would be more appropriate, plus it's also what was used for `samalba/dockerclient`.

calavera commented:
@tiborvass that name is totally provisional, until we move all that code to a dedicated repo.

I am not going to call it `dockerclient` because it doesn't follow that other library's conventions and it can be confusing and misleading.

PS: The repo is already created, but it's private for now :smile_cat:

calavera commented:
@tiborvass 

> I wonder if we shouldn't bring runconfig to the API

We should. I'd rather leave the configuration structs where they are for now, this is already big as it is. But we're working on consolidating those structs that are shared by client and server.

> I don't like that JSONConfig is a string

Fair enough, I'll make the change.

calavera committed:
Make the commit configuration to be a typed struct rather than accepting a string.

Signed-off-by: David Calavera <david.calavera@gmail.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "api_client_lib" git@github.com:calavera/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~53
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

tonistiigi commented:
I looked over most of the changes and played with this a bit. I think the API we provide should be in a bit higher level and easier to use for the third party apps, but I think there is no need to start going over these changes now. As a ported state of what our CLI needs this is a good starting point. But we should add a comment to the api package about this API being highly unstable atm.

I think we should try to get this in ASAP, to avoid heavy rebases and have plenty of time to test this before release. Everything doesn't need to be perfect in the first PR.

LGTM

tiborvass commented:
Thanks for the fix, LGTM

tiborvass commented:
userns had a flaky test I reported. Merging

tiborvass commented:
My thoughts go to all the people who now have to rebase

icecrime commented:
And to their families.

thaJeztah commented:
And to those having to ping "needs a rebase", lol

vdemeester commented:
:cat:

jhowardmsft commented:
@calavera Can you take a look at this change - it's broken Windows building on master....

calavera commented:
@jhowardmsft checking

calavera commented:
@jhowardmsft it looks like windows is still building with Go 1.4.2 while the rest is building with 1.5. Is that intentional?

The problem is that this function was added to 1.5:

https://golang.org/pkg/text/template/#Template.Option

We can go back to the workaround we had with 1.4.2. Although I'd rather not, the fix for 1.4.2 is to replace the current rawElementInspect implementation with this function:

https://github.com/docker/docker/blob/release/v1.9/api/client/inspect.go#L175-L203

duglin commented:
well, if we did go back to 1.4.2 that would fix the issues we're seeing with building on Power with gccgo where 1.5 isn't official available yet

ping @brahmaroutu

tianon commented:
https://github.com/docker/docker/blob/master/project/PACKAGERS.md#build-dependencies

We do still officially support compiling with Go 1.4, so we'll need to
figure out something (or decide whether that requirement can be bumped, but
it's sounding like that's not really plausible right at the moment).

calavera commented:
okay, I'll remove that call and go back to the ugly workaround :pensive: 

Thanks for checking!

LK4D4 commented:
@calavera You can use version build tags!

calavera commented:
@LK4D4 indeed. Doing that.

jhowardmsft commented:
@calavera Hang fire - I'll update the CI nodes, shouldn't take too long. I have already updated node 1 as I'm playing with Windows CI on it. @mikedougherty I assume you're OK me updating 2..9?

brahmaroutu commented:
gccgo master builds failed because of this. I think the best would be to add gccgo PR builds to catch go1.5 related breakage earlier? 
@tianon go1.5 support is on main branch, can we build gcc:6.0 image from master and use it for x86, Power and Z?

tianon commented:
It's OK, he's adding a shim for Go 1.4 support (which is still worthwhile
to keep for now IMO).

jhowardmsft commented:
We have to move to 1.5.2 on those machines anyway. I just accelerated the update. All nodes are updated with the exception of Jenkins-prs-3. @mikedougherty I sent you an email about that node - I was unable to logon to it, even after a restart through the portal.

WeiZhang555 commented:
Wow, this is really amazing!! :+1:

mikedougherty commented:
@jhowardmsft just got host 3 back up and running. sent you an email with details. thanks for the ping
